Dashboard v2 Preference Page: Interface Language section

andrea-sdl opened the pull request:
## Proposed Changes

* Introducing the new Preference section with the Interface language settings

### Demo of the changes

https://github.com/user-attachments/assets/9abbcb42-3893-4d9f-a46f-96b2ded1d9ff

**Example: Successful change notice**
<img width="2286" height="1792" alt="CleanShot 2025-09-05 at 15 31 08@2x" src="https://github.com/user-attachments/assets/8078d216-77b0-476a-b169-159f37d20a38" />

**Example: All available checkboxes are visible**
<img width="1634" height="1114" alt="CleanShot 2025-09-05 at 15 30 16@2x" src="https://github.com/user-attachments/assets/3e0babef-aa54-4817-b9f7-747b12a935f0" />

## Why are these changes being made?

* Porting features from V1

### Notes about the changes

The previous logic of changing the language was using a modal and a language picker. In this implementation, we're going to use the ComboBox control component to allow searching of languages. 

I reintroduced all the logic and checkboxes around empathy mode, in-app translation, and displaying interface in english.

The change also adds some basic unit tests to ensure that the logic behaves as we intend for non-translatable languages, as well as incomplete languages. 

There is one minor UI glitch, which is the alignment (and size) of the description for the in-page translator.
This is because the DataForm `description` field only accepts strings, and this is an actual component (since we need the link).

**Regarding validation:** The ComboboxControl doesn't have validations baked in, but since we do require a language to be set, I used the DataForm `isValid` to control the `allowReset` field of the ComboboxControl component, the language data also controls the status of the save button, marking it as disabled if no language has been selected.

## Testing Instructions

Compare https://wordpress.com/me/account (Interface language section) with http://calypso.localhost:3000/v2/me/preferences

Some useful languages to test/compare are
* English
* English (UK)
* Español de Chile
* Português

These languages will let you see the various checkbox, please note that the "in-page translator" toggle, is present outside of the modal, while the "Display Language in english" and the Empathy checkbox are available in the LanguagePicker modal.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

andrea-sdl committed:
exclude "@automattic/languages" from the no-restricted imports eslint rule

andrea-sdl committed:
add preferences menu to the "Me" section

andrea-sdl committed:
added an initial draft for the Preferences/Language Picker

andrea-sdl committed:
Merge branch 'trunk' into add/dashboard-v2-preference-page

andrea-sdl committed:
first iteration of a working example for the Preferences page

andrea-sdl committed:
Merge branch 'trunk' into add/dashboard-v2-preference-page

andrea-sdl committed:
add the preferences route

andrea-sdl committed:
add empathy mode

andrea-sdl committed:
Merge branch 'trunk' into add/dashboard-v2-preference-page

andrea-sdl committed:
add the community translator logic to show/hide the community translator message

andrea-sdl committed:
refactor: simplify language variant handling and conditionally show translator option

andrea-sdl committed:
add empathy mode field support for disabled status when using the default language

andrea-sdl committed:
disable submit button when it doesn't have a valid language selected

andrea-sdl committed:
use isVisible instead of conditionals to show fields.

andrea-sdl committed:
refactor: update language validation and add required field for interface language

andrea-sdl committed:
Merge branch 'trunk' into add/dashboard-v2-preference-page

andrea-sdl committed:
Move user settings preferences to API packages

andrea-sdl committed:
Add translation TODOs

oandregal commented on client/dashboard/me/preferences/index.tsx:
```diff
@@ -0,0 +1,281 @@
+			type: 'text',
+			isValid: {
+				required: true,
+			},
```
This field provides a custom `Edit` implementation, which means it needs to handle the `isValid` requirements itself — otherwise, this is not doing anything. See an example [implementation](https://github.com/WordPress/gutenberg/blob/2646370b28ffaf1a7048f688cbe13491ed6e0379/packages/dataviews/src/dataform-controls/text.tsx#L22) and it's also part of the storybook [examples](https://github.com/WordPress/gutenberg/blob/2646370b28ffaf1a7048f688cbe13491ed6e0379/packages/dataviews/src/components/dataform/stories/index.story.tsx#L369).

andrea-sdl replied:
Thank you for the feedback! I don't think we need complex validations here since, before, there was no validation that I can find (the field was required). Therefore, for now, I blocked the ability to reset the `ComboboxControl` and disabled the `Save` button in case there is no language selected. Thank you again for the good examples, they were useful in understanding how things work.

andrea-sdl committed:
add the `isValid` field to indicate that the combobox languagevalue is required.

andrea-sdl committed:
Refactor user preferences to use api-core and add translator option

- Remove local me-user-preferences data module
- Use UserSettingsPreferences from @automattic/api-core
- Add "enable_translator" to preferences and API mutator
- Move languages data to preferences directory
- Update language form to support incomplete locale fallback and
  translator option

andrea-sdl committed:
Remove description for empathy mode in UserSettingsLanguageForm

andrea-sdl committed:
Merge branch 'trunk' into add/dashboard-v2-preference-page

andrea-sdl committed:
add better comments to explain why we're overriding the serverData

andrea-sdl committed:
move language functions into languages.ts file

andrea-sdl committed:
fix: correct logic for default language selection in UserSettingsLanguageForm

andrea-sdl committed:
feat: add success notice and URL handling after language setting save

andrea-sdl committed:
fix: remove unnecessary eslint-disable comment for restricted imports

andrea-sdl committed:
move the language form into a separate folder

andrea-sdl committed:
add languages test

andrea-sdl committed:
test: add comprehensive tests for language utilities

andrea-sdl committed:
test: add unit tests for PreferencesLanguageForm component

andrea-sdl committed:
refactor: remove TODO comment for adding tests in PreferencesLanguageForm

andrea-sdl committed:
require the language

andrea-sdl committed:
fix: update external link markup in community translator message

andrea-sdl committed:
feat: import UserSettingsPreferences from api-core in languages test file

abdullah-kasim commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+	 * as such, we're overriding the language with the locale_variant, if present, this saves us from checking the data and allows us to
+	 * trust that the 'language' field contains a locale from the languages.
+	 */
+	if ( serverData?.locale_variant && serverData.locale_variant !== '' ) {
```
nit: The second conditional never returns false. `serverData.locale_variant` is already falsey if it's an empty string.

andrea-sdl replied:
fixed in a91193e39c050cabc6dc14fcbfcfefc513045d31

abdullah-kasim commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+		} );
+	};
+
+	const isDefaultLanguageSelected = !! data.language && isDefaultLocale( data.language );
```
```suggestion
	const isDefaultLanguageSelected = Boolean( data.language ) && isDefaultLocale( data.language );
```

Feels more readable to me - WDYT? But this may be because I came from VIP and VIP's codebase prefers `Boolean()`. Though if the codebase prefers `!!`, I'm good with that.

andrea-sdl replied:
@abdullah-kasim I like your approach more, but then TS would still require a check for that field existance since `isDefaultLocale` doesn't accept undefined (only `string` or `null`)

abdullah-kasim replied:
```suggestion
	const isDefaultLanguageSelected = Boolean( data.language && isDefaultLocale( data.language ) );
```

Aw, that's unfortunate. I'd wrap everything up with `Boolean` in that case, or do `isDefaultLocale( data.language || '' )` - even though I might be living dangerously in that manner.

Still, just a nit for me since this doesn't break a linting rule 👍

andrea-sdl replied:
Yeah, for now I'll keep it this way as I've seen `!!` being used multiple times in the codebase.

abdullah-kasim commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+			return serverData[ key as keyof UserSettingsPreferences ] !== value;
+		} );
+
+	const hasValidLanguage = !! data?.language && data.language !== '';
```
```suggestion
	const hasValidLanguage = !! data?.language;
```

nit: similar to the above comment - the second condition will always return true because an empty string is already falsey and hence, JS will skip the subsequent conditionals.

andrea-sdl replied:
fixed in 0162e6358dfac350365d9e202ea0bf8f75a64ea2

abdullah-kasim commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+	const { data: serverData } = useQuery( userSettingsPreferencesQuery() );
+	const [ localData, setLocalData ] = useState< Partial< UserSettingsPreferences > | undefined >();
+	const mutation = useMutation( userSettingsPreferencesMutation() );
+	const [ savingData, setSavingData ] = useState<
+		Partial< UserSettingsPreferences > | undefined
+	>();
```
No action needed but just wanted to better my understanding of this codebase - what's the intent of `savingData`? At first glance, it doesn't look necessary.

arthur791004 replied:
Agreed. It seems the `localeData` (or call it `formData`) is enough 🤔

andrea-sdl replied:
Yeah, good catch. It's was probably a leftover of my initial code, changed in 9979dda0ee365dd07d295bb0452297a48d3c0d60

abdullah-kasim commented on client/dashboard/me/preferences/index.tsx:
```diff
@@ -0,0 +1,12 @@
+import PageLayout from '../../components/page-layout';
+import PreferencesLanguageForm from '../preferences-language';
+
+export default function Preferences() {
+	return (
+		<PageLayout size="small" header={ <PageHeader title={ __( 'Preferences' ) } /> }>
+			<PreferencesLanguageForm />
+		</PageLayout>
+	);
+}
```
❤️ on the folder structuring!

andrea-sdl replied:
I got inspired by you ❤️

abdullah-kasim reviewed (commented):
Did a quick review on the components. I'm not too confident about fully reviewing the tests and the code in the `packages` folder yet.

arthur791004 commented on packages/api-queries/src/me-user-settings-preferences.ts:
```diff
@@ -0,0 +1,20 @@
+		onSuccess: ( newData ) => {
+			queryClient.setQueryData(
+				userSettingsPreferencesQuery().queryKey,
+				( oldData ) => oldData && { ...oldData, ...newData }
```
Do we really need to check the `oldData` 🤔

andrea-sdl replied:
yeah, we don't. Removed.

arthur791004 commented on packages/api-core/src/me-user-settings-preferences/mutators.ts:
```diff
@@ -0,0 +1,20 @@
+			delete data[ key as keyof UserSettingsPreferences ];
+		}
+	}
+	return await wpcom.req.post( '/me/settings', data );
```
Not a blocker: Would it be better to create a new object instead of deleting the key from the input?

```js
const payload = Object.fromEntries(
    saveableKeys
      .filter(key => key in obj)
      .map(key => [key, obj[key]])
  );
```

andrea-sdl replied:
Nice, I like it. I did it here (with some extra type checks) 690f06ff2878815d067d433f6ac0624593beb0e5

arthur791004 commented on client/dashboard/me/preferences-language/languages.ts:
```diff
@@ -0,0 +1,50 @@
+ * Adapted from https://github.com/Automattic/wp-calypso/blob/fbeb9c37266e2bfac7af881b1672a9f6d72a0670/client/me/account/main.jsx#L299
+ * In this case the data.language is the locale variant if we're using one, so we can skip the "isLocaleVariant checks and see if the locale can be translated or not"
+ */
+export const shouldDisplayCommunityTranslator = ( data: UserSettingsPreferences ): boolean => {
```
How about just accepting the `locale` as input instead of all preferences?

andrea-sdl replied:
same answer as https://github.com/Automattic/wp-calypso/pull/105311#discussion_r2321451246

andrea-sdl replied:
fixed in 09a7212334041cbe1efb1295dfd022b3e7581b5e

arthur791004 commented on client/dashboard/me/preferences-language/languages.ts:
```diff
@@ -0,0 +1,50 @@
+ * Get the locale variant if the parentLangSlug is present
+ */
+export const getLocaleVariantOrLanguage = (
+	data: UserSettingsPreferences
```
How about just accepting the `language` as input instead of all preferences?

andrea-sdl replied:
That's a good idea. Initially the function was more entertwined with the rest and more complex, but it simplified once I added the https://github.com/Automattic/wp-calypso/blob/9979dda0ee365dd07d295bb0452297a48d3c0d60/client/dashboard/me/preferences-language/index.tsx#L91 so it makes absolute sense to simplify it.

andrea-sdl replied:
fixed in 09a7212334041cbe1efb1295dfd022b3e7581b5e

arthur791004 commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+	];
+
+	return (
+		<>
```
We don't need outer fragment 🙂

andrea-sdl replied:
Removed, good catch!

arthur791004 commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+								<Button
+									variant="primary"
+									type="submit"
+									className="language-preferences-form__submit"
```
We don't need this class

andrea-sdl replied:
Removed (probably me copy/pasting or testing things in the beginning)

arthur791004 commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+import { UserSettingsPreferences } from '@automattic/api-core';
```
It would be better to just import type like

```js
import type { UserSettingsPreferences } from '@automattic/api-core';
```

arthur791004 commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+} from './languages';
+import type { Field } from '@wordpress/dataviews';
+
+const thanksToCommunityTranslator = ( data: UserSettingsPreferences ) => {
```
Should it be a component?

andrea-sdl replied:
I was initially unsure about this, as it didn't felt like very reusable, but it definitely helps with readability and clarity. Added the component in cd784e5cde37f56952e803a5f69eb697bc255bfa

arthur791004 commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+							} }
+							placeholder={ __( 'Select a language' ) }
+							options={ field.elements || [] }
+							allowReset={ false } // a language is required so we're not allowing to reset it and have an empty state.
```
It would be better to set `__next40pxDefaultSize`

andrea-sdl replied:
Thank you for flagging! Should we use __next40pxDefaultSize whenever possible?

arthur791004 commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+					<>
+						<CheckboxControl
+							checked={ field.getValue( { item: data } ) }
+							label={ field.label }
```
Since we have custom Edit here, we can simply pass description to `help` prop of the `CheckboxControl` 🙂

arthur791004 commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,310 @@
+							{ __(
+								'This is the language of the interface you see across WordPress.com as a whole.'
+							) }
+							{ data && thanksToCommunityTranslator( data ) }
```
We can use the `help` prop of the ComboboxControl to display it

andrea-sdl replied:
done in 51c22bae134e0d1fb16514b1a953d3b99cd46996

andrea-sdl committed:
remove useless extra check for the locale_variant

andrea-sdl committed:
remove extra content check since it's already catched by the previous one

andrea-sdl committed:
remove savingData since it's not used anymore and rename localData to formData

andrea-sdl committed:
rely only on the locale for the getLocaleVariantOrLanguage and shouldDisplayCommunityTranslator instead of passing the full settings object

andrea-sdl committed:
add __next40pxDefaultSize to the ComboboxControl

andrea-sdl committed:
use the help property instead of additional text components to rely on default DataForm logic

andrea-sdl committed:
remove unused import

andrea-sdl committed:
refactor: move community translator logic to a separate component

andrea-sdl committed:
fix: import UserSettingsPreferences type correctly

andrea-sdl committed:
remove not-needed outer fragment

andrea-sdl committed:
remove unused class

andrea-sdl committed:
remove oldData since it's not needed

andrea-sdl committed:
use filter (with better typing) to filter the fields to save instead of deleting the keys from the object

andrea-sdl committed:
use a parameter name that can be used by other components if needed

andrea-sdl committed:
reintroduce the old data, to prevent an UI glitch when saving

andrea-sdl committed:
remove oldData checking

andrea-sdl committed:
rename querykey to user-settings-preferences

andrea-sdl committed:
fix caching problems when refreshing the page on save

abdullah-kasim commented on client/dashboard/me/preferences-language/index.tsx:
```diff
@@ -0,0 +1,258 @@
+
+	const hasValidLanguage = !! data?.language;
+	const canSubmit = ! isSaving && isDirty && hasValidLanguage;
+	const languageForm = {
+		type: 'regular' as const,
+		labelPosition: 'top' as const,
+		fields: [
+			{
+				id: 'interfaceLanguage',
+				label: __( 'Language' ),
+				children: [
+					'language',
+					'use_fallback_for_incomplete_languages',
+					'enable_translator',
+					'i18n_empathy_mode',
+				],
+			},
+		],
+	};
```
```suggestion
	const languageForm: React.ComponentProps<DataForm< UserSettingsPreferences >>['form'] = {
		type: 'regular' as const,
		labelPosition: 'top' as const,
		fields: [
			{
				id: 'interfaceLanguage',
				label: __( 'Language' ),
				children: [
					'language',
					'use_fallback_for_incomplete_languages',
					'enable_translator',
					'i18n_empathy_mode',
				],
			},
		],
	};
```

Try something similar to this (syntax might be slightly wrong). I do this frequently to ensure that I get some type-checking and code-completion at this particular spot rather than down below at the component.

andrea-sdl replied:
added in https://github.com/Automattic/wp-calypso/pull/105311/commits/95a7b6cf75879045084c251a00bf180864655f6c, I've used a shorter notation that I found in another component that, in turn, helped find a type error :) thank you!

abdullah-kasim reviewed (commented):
LGTM!

I like how proactive you are whenever you see that an improvement can be made. Like, changing the variable name from localeData to formData makes a big improvement in readability!

andrea-sdl committed:
add typing to languageForm

arthur791004 reviewed (approved):
Looks good 👍
